stats.go

```diff
@@ -193,27 +207,233 @@ func (t *timer) snapshot() []time.Duration {
+	name string,
+	tags map[string]string,
+	buckets Buckets,
+	bucketLowerBound,
+	bucketUpperBound time.Duration,
+	samples int64,
+) {
+}
+
 func (r *timerNoReporterSink) Capabilities() Capabilities {
-	return capabilitiesReportingTagging
+	return capabilitiesReportingTaggingNoHistograms
 }
+
 func (r *timerNoReporterSink) Flush() {
 }
 
-// NullStatsReporter is an implementatin of StatsReporter than simply does nothing.
+type histogram struct {
+	htype            histogramType
+	name             string
+	tags             map[string]string
+	reporter         StatsReporter
+	specification    Buckets
+	buckets          []histogramBucket
+	lookupByValue    []float64
+	lookupByDuration []int
+}
+
+type histogramType int
+
+const (
+	valueHistogramType histogramType = iota
+	durationHistogramType
+)
+
+func newHistogram(
+	name string,
+	tags map[string]string,
+	reporter StatsReporter,
+	buckets Buckets,
+	cachedHistogram CachedHistogram,
+) *histogram {
+	htype := valueHistogramType
+	if _, ok := buckets.(DurationBuckets); ok {
+		htype = durationHistogramType
+	}
+
+	pairs := BucketPairs(buckets)
+
+	h := &histogram{
+		htype:            htype,
+		name:             name,
+		tags:             tags,
+		reporter:         reporter,
+		specification:    buckets,
+		buckets:          make([]histogramBucket, 0, len(pairs)),
+		lookupByValue:    make([]float64, 0, len(pairs)),
+		lookupByDuration: make([]int, 0, len(pairs)),
+	}
+
+	for _, pair := range pairs {
+		h.addBucket(newHistogramBucket(h,
+			pair.LowerBoundValue(), pair.UpperBoundValue(),
+			pair.LowerBoundDuration(), pair.UpperBoundDuration(),
+			cachedHistogram))
+	}
+
+	return h
+}
+
+func (h *histogram) addBucket(b histogramBucket) {
+	h.buckets = append(h.buckets, b)
+	h.lookupByValue = append(h.lookupByValue, b.valueUpperBound)
+	h.lookupByDuration = append(h.lookupByDuration, int(b.durationUpperBound))
+}
+
+func (h *histogram) report(name string, tags map[string]string, r StatsReporter) {
+	for i := range h.buckets {
+		samples := h.buckets[i].samples.value()
+		if samples == 0 {
+			continue
+		}
+		switch h.htype {
+		case valueHistogramType:
+			r.ReportHistogramValueSamples(name, tags, h.specification,
+				h.buckets[i].valueLowerBound, h.buckets[i].valueUpperBound,
+				samples)
+		case durationHistogramType:
+			r.ReportHistogramDurationSamples(name, tags, h.specification,
+				h.buckets[i].durationLowerBound, h.buckets[i].durationUpperBound,
+				samples)
+		}
+	}
+}
+
+func (h *histogram) cachedReport() {
+	for i := range h.buckets {
+		samples := h.buckets[i].samples.value()
+		if samples == 0 {
+			continue
+		}
+		switch h.htype {
+		case valueHistogramType:
+			h.buckets[i].cachedValueBucket.ReportSamples(samples)
+		case durationHistogramType:
+			h.buckets[i].cachedDurationBucket.ReportSamples(samples)
+		}
+	}
+}
+
+func (h *histogram) RecordValue(value float64) {
+	idx := sort.SearchFloat64s(h.lookupByValue, value)
```

Comment:
Need a comment here documenting the lowest value in `lookupByValue` is the min float64 and the highest value in `lookupByValue` is max float64